rfc/20240513-static-evaluation-providers.md

```diff
@@ -116,206 +112,394 @@ locals {
+
+The `configs.Provider` struct representing each configured `provider` block will now contain new field that contains the static mapping from "provider instance key" to "provider repetition data". In practice, this is a map of `addrs.InstanceKey` to `instances.RepetitionData`.  This map can be built using the StaticContext available in [configs.NewModule()](https://github.com/opentofu/opentofu/blob/290fbd66d3f95d3fa413534c4d5e14ef7d95ea2e/internal/configs/module.go#L122) as defined in the Static Evaluation RFC.
+
+At the end of the `configs.NewModule` constructor, all provider configurations that contain a for_each or count will have their new "instance mapping" field set (or an error message produced). This should not change the majority of provider validation logic in the configuration package as the name/type/alias information *has not changed*.
+
+##### Provider Node Execution
+
+Each "provider configuration block" is turned into a "provider node" (NodeApplyableProvider) within the tofu graph. When a provider node is executed, it starts up and configures a `providers.Interface` of the corresponding provider. This "interface" is stored in the `tofu.EvalContext` and is available for use by other nodes via the context (referenced below).
+
+`NodeApplyableProvider` now effectively represents all `provider.Interface`s of a particular `provider` configuration block at once, and its "execute" step now involves a loop performing for each instance similar behavior that was previously always singleton:
+1. Evaluate the configuration using the appropriate `instances.RepetitionData`, thereby causing `each.key` and `each.value` to be available with instance-specific values. The result is a `cty.Value` representing an object of the type implied by the provider's configuration schema.
+2. Create a child process running the appropriate provider plugin executable, creating a running provider interface.
+3. Call the `ConfigureProvider` function from the provider protocol on the new provider interface, passing it the `cty.Value` representing its instance-specific configuration.
+4. Save the provider interface in the shared `tofu.EvalContext` for use by downstream resource nodes.
+
+A special case exists here for `tofu validate`. The validate codepath does not configure any providers or deal with "expansion" (for_each/count). A validate graph walk should only initialize and register one `providers.Interface`.  It can still however validate the configuration for each set of provider instance data.
```

Comment:
Can you say some more about what the last sentence here implies? It's not clear to me what exactly "the configuration" refers to, and what "each set of provider instance data" refers to in the validation walk where there are not typically any individual instances. :thinking:

Reply:
[f1de6ba](https://github.com/opentofu/opentofu/pull/2115/commits/f1de6ba99653bbe56da253b84777bec7aa6c7c18)